### PR TITLE
TOML declarative config for modules 3/4

### PR DESCRIPTION
**Pull request by chrzaszcz**
Modules covered in this PR:
- [x] mod_muc
- [x] mod_muc_light
- [x] mod_muc_log
- [x] mod_offline
- [x] mod_ping
- [x] mod_privacy
- [x] mod_private 
- [x] mod_pubsub
- [x] mod_push_service_mongoosepush

Checklist for each module:
- Write the new spec in the module as a callback function: `config_spec/0`. Some cases need reworking the options, e.g. when the same value can be a primitive (e.g. string or atom) or a nested section/list
- Add the module to `mongoose_config_spec:all_modules/0` and `HAS_NO_SPEC`
- Double check with the docs: `https://esl.github.io/MongooseDocs` that the options and their format, validator etc are correct
- Remove `module_opt` and `handler` clauses from `mongoose_config_parser_toml` as well as module-specific functions for nested options
- Remove `validate` clauses from `mongoose_config_validator_toml`
- Make sure this passes locally: `./rebar3 ct --suite=config_parser_SUITE` before pushing anything, also check `./rebar3 dialyzer`
- Check test coverage after pushing, there might be some options not covered by tests, you can do it by looking at the test cases as well
- Regarding tests, it's good to simplify them and fix indentation/maximum line length of 100 characters, see the change for `mod_adhoc` or `mod_auth_token`.
- Sanity check: grep for the module name in `src/config`

### CODE REVIEW

File: src/pubsub/mod_pubsub.erl [hunk 1]
```diff
@@ -240,6 +245,78 @@ stop(Host) ->
...
+    #section{
+       items = #{<<"namespace">> => #option{type = binary,
+                                            validate = non_empty},
+                 <<"node">> => #option{type = binary,
+                                       validate = non_empty}},
+       required = all,
+       process = fun ?MODULE:process_pep_mapping/1
+      }.
+
+default_node_config_spec() ->
```
Comment (janciesla8818) [inline src/pubsub/mod_pubsub.erl:286]: I've added a backlog documentation item to add a list of node options that can be configured.

**Review by janciesla8818 — APPROVED** [review]
Good job 💪 all looks good. Thanks for the effort and the attention to detail.

File: test/config_parser_SUITE.erl [hunk 2]
```diff
@@ -1997,494 +2003,446 @@ mam_failing_riak_cases(T) ->
...
+                                              #{<<"pep_mapping">> => Opts}}} end,
+    M = fun(Cfg) -> modopts(mod_pubsub, [{pep_mapping, Cfg}]) end,
+    RequiredOpts = #{<<"namespace">> => <<"urn:xmpp:microblog:0">>,
+                     <<"node">> => <<"mb">>},
+    ?eqf(M([{<<"urn:xmpp:microblog:0">>, <<"mb">>}]),
+         T([RequiredOpts])),
+    [?errf(T([maps:remove(Key, RequiredOpts)])) || Key <- maps:keys(RequiredOpts)],
+    [?errf(T([RequiredOpts#{Key => <<>>}])) || Key <- maps:keys(RequiredOpts)].
+
+mod_pubsub_default_node_config(_Config) ->
```
Comment (leszke) [inline test/config_parser_SUITE.erl:2381]: `max_items` option has not been tested.

File: src/pubsub/mod_pubsub.erl (hunk 1 above)
Comment (chrzaszcz) [reply src/pubsub/mod_pubsub.erl:286]: Good, thanks

File: src/mod_muc_log.erl [hunk 3]
```diff
@@ -72,7 +77,7 @@
                    dir_type     :: dir_type(),
                    dir_name     :: dir_name(),
                    file_format  :: file_format(),
-                   css_file     :: file:filename() | false,
+                   css_file     :: binary() | false,
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/mod_muc_log.erl:80]: It defaults to false in the code: `CSSFile = gen_mod:get_opt(cssfile, Opts, false),`

File: src/pubsub/mod_pubsub.erl [hunk 4]
```diff
@@ -240,6 +245,78 @@ stop(Host) ->
...
+                 <<"host">> => #option{type = string,
+                                       validate = domain_template},
+                 <<"backend">> => #option{type = atom,
+                                          validate = {module, mod_pubsub_db}},
+                 <<"access_createnode">> => #option{type = atom,
+                                                    validate = access_rule},
+                 <<"max_items_node">> => #option{type = integer,
+                                                 validate = non_negative},
+                 <<"max_subscriptions_node">> => #option{type = integer,
+                                                         validate = non_negative},
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/pubsub/mod_pubsub.erl:261]: It defaults to `undefined` if the option is missing:    ` MaxSubsNode = gen_mod:get_opt(max_subscriptions_node, Opts,
                                  fun(A) when is_integer(A) andalso A >= 0 -> A end, undefined),`

File: test/config_parser_SUITE.erl (hunk 2 above)
Comment (chrzaszcz) [reply test/config_parser_SUITE.erl:2381]: Good catch!

**Review by leszke — APPROVED** [review]

File: src/mod_muc.erl [hunk 5]
```diff
@@ -164,6 +167,108 @@ stop(Host) ->
...
+                 <<"maygetmemberlist">> => #list{items = #option{type = atom,
+                                                                 validate = non_empty}},
+                 <<"affiliations">> => #list{items = default_room_affiliations_spec()},
+                 <<"subject">> => #option{type = binary},
+                 <<"subject_author">> => #option{type = binary}
+                },
+       format = {kv, default_room_options}
+      }.
+
+default_room_affiliations_spec() ->
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/mod_muc.erl:254]: Could be, I just avoided too much verbosity.

File: src/muc_light/mod_muc_light.erl [hunk 6]
```diff
@@ -192,6 +196,54 @@ stop(Host) ->
...
+                 <<"max_occupants">> => #option{type = int_or_infinity,
+                                                validate = positive},
+                 <<"rooms_per_page">> => #option{type = int_or_infinity,
+                                                 validate = positive},
+                 <<"rooms_in_rosters">> => #option{type = boolean},
+                 <<"config_schema">> => #list{items = config_schema_spec()}
+                }
+      }.
+
+config_schema_spec() ->
```
[the remark this answers is not in the document]
Comment (chrzaszcz) [reply src/muc_light/mod_muc_light.erl:222]: The same, I just did not want too much config's there, to keep some sanity.

**Pull request merged**